### PR TITLE
Implement Reflect for Box<dyn Reflect>

### LINKED ISSUE #3392: Implement Reflect for Box<dyn Reflect>
## What problem does this solve or what need does it fill?

Boxed Reflect trait objects should themselves be reflectable

## What solution would you like?

Implement `Reflect` for `Box<dyn Reflect>`.

## What alternative(s) have you considered?

None.

## Additional context

Working snippet from @Davier:

```rust
unsafe impl Reflect for Box<dyn Reflect> {
    fn type_name(&self) -> &str {
        self.deref().type_name()
    }

    fn any(&self) -> &dyn Any {
        self.deref().any()
    }

    fn any_mut(&mut self) -> &mut dyn Any {
        self.deref_mut().any_mut()
    }

    fn apply(&mut self, value: &dyn Reflect) {
        self.deref_mut().apply(value)
    }

    fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>> {
        self.deref_mut().set(value)
    }

    fn reflect_ref(&self) -> ReflectRef {
        self.deref().reflect_ref()
    }

    fn reflect_mut(&mut self) -> ReflectMut {
        self.deref_mut().reflect_mut()
    }

    fn clone_value(&self) -> Box<dyn Reflect> {
        self.deref().clone_value()
    }

    fn reflect_hash(&self) -> Option<u64> {
        self.deref().reflect_hash()
    }

    fn reflect_partial_eq(&self, value: &dyn Reflect) -> Option<bool> {
        self.deref().reflect_partial_eq(value)
    }

    fn serializable(&self) -> Option<Serializable> {
        self.deref().serializable()
    }
}
```

### PR DESCRIPTION
**Pull request by franciscoaguirre**
# Objective

Closes #3392 

## Solution

Used code snippet from @Davier and made a few changes on `DynamicList` for the type checker to work.

### CODE REVIEW

File: crates/bevy_reflect/src/list.rs [hunk 1]
```diff
@@ -46,11 +46,11 @@ impl DynamicList {
 
 impl List for DynamicList {
     fn get(&self, index: usize) -> Option<&dyn Reflect> {
-        self.values.get(index).map(|value| &**value)
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/list.rs:49]: Interesting; why were these changes required? 

They certainly seem reasonable, but I'm not familiar enough with the original code to really grasp why it was needed in the first place.

Comment (alice-i-cecile) [reply crates/bevy_reflect/src/list.rs:49]: It looks like we have reasonable (if basic) tests for this type, which is reassuring, as they continue to pass :)

**Comment by alice-i-cecile** [comment]
Awesome, thank you. This is a nice addition, and your test looks nice :)

File: crates/bevy_reflect/src/list.rs (hunk 1 above)
Comment (franciscoaguirre) [reply crates/bevy_reflect/src/list.rs:49]: I think this is why it was done like this previously:
`get` called on `vec` results in an `Option<&Box<dyn Reflect>>`, in order to return an `Option<&dyn Reflect>` it uses `&**` (dereferences it twice to get to the trait object and then gets a reference to it).

Now, because of the new trait:
`Option<&Box<dyn Reflect>>`, the return value from `get`, already acts as `Option<&dyn Reflect>` (the return value), which already works.
The error I got was that `dyn Reflect` can't be dereferenced, so I don't know why it's acting as a `dyn Reflect` and not a `Box`, which is the actual type.
I believe this should work anyway since the Box will forward all reflect calls.

Comment (alice-i-cecile) [reply crates/bevy_reflect/src/list.rs:49]: > The error I got was that dyn Reflect can't be dereferenced, so I don't know why it's acting as a dyn Reflect and not a Box, which is the actual type.

Presumably it was dispatching directly to the `Box` type now that it implements reflect?

Awesome, your explanation makes sense. This is a nice improvement to the readability of that code, so I'm happy with it.

**Review by alice-i-cecile — APPROVED** [review]

**Review by Davier — APPROVED** [review]
Looks good to me!
A possible improvement would be to show its usage in the `reflection.rs` example, perhaps by adding a field of type `Box<dyn Reflect>` in `Foo`?

**Review by alice-i-cecile — APPROVED** [review]

**Review by MrGVSV — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@cart do your concerns from #2505 apply here?

**Comment by cart** [comment]
Yup "boxing boxes" is definitely a concern here, although the scope of the problem is different because in core bevy apis we aren't implicitly boxing `T: Reflect` parameters that might later be retrieved and passed back in (at least to my knowledge).

However `Box<dyn Reflect>` is used/returned in a number of places (`Reflect::clone_value()`, scenes, etc), so this feels risky to me.

Thaaaat being said, having reflection work for "everything", including Boxed reflect types, does _seem_ like it could be valuable. Can we outline specific "real world" scenarios that need this (that can't be resolved in other ways?). So far I've only seen that a user wants to do this, which isn't enough context to evaluate the need here.

**Comment by djeedai** [comment]
Note that any `impl Trait for Box<dyn Trait>` opens the door to double-boxing (nested boxing) and especially to silently doing so by mistake. Coincidentally we're hitting this in 🍃Bevy Tweening with djeedai/bevy_tweening#27. See discussion highlighting a few important points:
1. even `rustlang/futures-rs` struggles with this, it's a hard issue, that trait specialization will solve once it's stable, but until then is hard to fix.
2. you don't need to not have that impl in the current crate for the problem to be avoided with any public trait, because any crate could add that impl for you, and your back to square one.
3. incidentally 2. makes it harder to work around double boxing.
4. there's some runtime workaround via `Any::try_downcast()` that was suggested to me on the Rust Discord, but it's a runtime check not a build time guarantee.

Given all of that I think it's reasonable to still add the impl, and thoroughly document places where nested boxing might occur by mistake. And consider adding that runtime check as a guard.

**Comment by djeedai** [comment]
Here's the runtime workaround : https://discord.com/channels/442252698964721669/443150878111694848/981137955773095936 (I didn't try it yet)

> Use this function:
```rust
fn try_downcast<T, K>(k: K) -> Result<T, K>
where
    T: 'static,
    K: 'static,
{
    let mut k = Some(k);
    if let Some(k) = <dyn std::any::Any>::downcast_mut::<Option<T>>(&mut k) {
        Ok(k.take().unwrap())
    } else {
        Err(k.unwrap())
    }
}
```
> like so: 
```rust
fn box_trait<T: 'static + Trait>(val: T) -> Box<dyn Trait> {
    try_downcast::<Box<dyn Trait>, _>(val)
        .unwrap_or_else(|val| Box::new(val))
}
```

**Comment by PROMETHIA-27** [comment]
as a real-world example; working on my editor I've stumbled across this. I want to send messages over network between the editor and the game client, and those messages are serialized into yaml (could be anything, picked arbitrarily) via reflection. In order to send reflected types, I sometimes need to send a Box<dyn Reflect> (serialized components), but it's not Reflect so I ended up implementing a ReflectObject newtype around it. This PR would allow me to throw that (likely somewhat buggy) code out and reduce annoying boilerplate/restrictions

File: crates/bevy_reflect/src/reflect.rs [hunk 2]
```diff
@@ -86,3 +90,49 @@ impl dyn Reflect {
         self.any_mut().downcast_mut::<T>()
     }
 }
+
+unsafe impl Reflect for Box<dyn Reflect> {
+    fn type_name(&self) -> &str {
+        self.deref().type_name()
```
Comment (PROMETHIA-27) [inline crates/bevy_reflect/src/reflect.rs:96]: why is `.deref()` necessary here? Clarity?

File: crates/bevy_reflect/src/reflect.rs [hunk 3]
```diff
@@ -86,3 +90,49 @@ impl dyn Reflect {
...
+    }
+
+    fn any(&self) -> &dyn Any {
+        self.deref().any()
```
Comment (PROMETHIA-27) [inline crates/bevy_reflect/src/reflect.rs:100]: Is this safe/correct? The unsafe contract for `impl Reflect` is that `any` and `any_mut` return the `self` value passed in. I could see this being correct but I could also see an edge case like boxing the `Box<dyn Reflect>` and using `downcast` on that being UB.

**Review by PROMETHIA-27 — COMMENTED** [review]
Looks decent, but I have a few questions:

File: crates/bevy_reflect/src/list.rs (hunk 1 above)
Comment (SkiFire13) [reply crates/bevy_reflect/src/list.rs:49]: I believe this should work, however I'm pretty sure this will result in double dynamic dispatch: the first dynamic call will call `Box<dyn Reflect>`'s method, which in turn will do a dynamic call on its inner value. Maybe this should need an `.as_deref()`?

File: crates/bevy_reflect/src/reflect.rs (hunk 3 above)
Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/reflect.rs:100]: Some solutions I can think of:
- Just return `self` instead of `self.deref().any()`
- Add `any_box` to `Reflect` which takes `self` and returns `Box<dyn Any>`, and then remove the unsafe in `downcast`, which should also make Reflect entirely safe to implement
- Move downcast to be a `Reflect` method instead of an impl on `dyn Reflect`?

**Comment by MrGVSV** [comment]
@franciscoaguirre Would you be able to rebase this off `main`?

**Comment by MrGVSV** [comment]
Adding the `S-Adopt-Me` tag. 

@franciscoaguirre let us know if you'd rather continue working on this yourself and I'll remove it! :)

**Comment by franciscoaguirre** [comment]
@MrGVSV Yeah I can work on this. There was so much debate that I got lost :joy: and I wasn't having much time. I have more time now so I can get it done

**Comment by rdrpenguin04** [comment]
@franciscoaguirre Are you still intending to work on this?

**Comment by franciscoaguirre** [comment]
@rdrpenguin04 I've been heavily caught in finishing university. Now that I'll soon graduate I'll have time for open source and I'd love to contribute to Bevy. That said, I can only work on it in January, since the holidays are coming up. I've seen people requesting this PR to be merged so if you need it before holidays I don't mind you taking the torch. If not, I can work on it in January :)

**Comment by franciscoaguirre** [comment]
@rdrpenguin04 I got around to rebasing and updating the code with the latest on main :smile: 
@PROMETHIA-27 Is this PR okay now with your change to the `Reflect` trait?
@alice-i-cecile Is this issue still relevant?

**Comment by djeedai** [comment]
@franciscoaguirre FYI you have at least one test panicking, looks like it's directly related to your change.

**Comment by franciscoaguirre** [comment]
Yep, fixed it.

The thing is, I was implementing `DynamicTypePath` on `Box<dyn Reflect>` which lets me use self to forward the call to the inner value, but I needed to implement TypePath for the type and having both implementations resulted in conflicts.
I think it's more useful to get the type of the inner value, maybe not? Is there a way I can accomplish that?

**Comment by djeedai** [comment]
I have no idea, I don't have enough domain knowledge to answer whether we should implement the dynamic version or not. My gut feeling is that `TypePath` would be more useful, but I really don't have a backing argument sorry. Maybe @alice-i-cecile or @MrGVSV can help here?

**Comment by JMS55** [comment]
Maintainers: What are we doing with this PR? It's been S-Ready-For-Final-Review for a while now

**Comment by alice-i-cecile** [comment]
Deferring to the Reflection SMEs. I'll bring it up on Discord.

**Comment by MrGVSV** [comment]
> Maintainers: What are we doing with this PR? It's been S-Ready-For-Final-Review for a while now

Sorry to everyone that's been waiting on this :/

I'm currently looking into a possible alternative approach that aims to avoid the double-boxing issue by going through remote reflection. However, if that doesn't work out or it runs into the same problems, then we'll probably move forward with this PR.

I'll keep experimenting and return with details soon!

Either way, I'd like to get something in for `Box<dyn Reflect>` and `Box<dyn PartialReflect>` for 0.15.

**Comment by MrGVSV** [comment]
I've just pushed #14776 if anyone wants to give it a look!

The TL;DR is that we can avoid the risk of double-boxing by adding a separate `ReflectBox` type (that users can't construct—otherwise we'd have the same problem) and using remote reflection to reflect it:

```rust
use bevy_reflect::boxed::ReflectBox;

#[derive(Reflect)]
#[reflect(from_reflect = false)]
struct Player {
    #[reflect(remote = ReflectBox<dyn Reflect>)]
    weapon: Box<dyn Reflect>,
}
```

The main consideration is that this `ReflectBox` approach comes with many limitations. Some are the same as this PR (i.e. doesn't work with `FromReflect`, can't be used in enums), but some are unique to remote reflection (i.e. difficulty using `ReflectBox` as a type parameter).

So the question becomes, do we go for the more flexible PR (this one) or the more conservative PR (#14776)?

I'm leaning towards starting out conservative to see how far that can take us. We can then revisit this PR in the future if necessary.

However, I'm more than happy to hear alternative perspectives or even other ideas!

---

Also, I wanted to point out that another cool benefit to `ReflectBox` is its ability to work with any type `T` that can be converted to a `dyn PartialReflect` (using a couple traits). This includes other trait objects like `dyn MyCoolTrait`.

If we do end up going with this PR, it might be worthwhile trying to get that behavior to work as well (possibly using a similar approach as `ReflectBox`). Either in this PR or a followup one.

**Comment by alice-i-cecile** [comment]
Closing in favor of #14776.
